### ISSUE TITLE
improve error handling for auth issues: "Internal Server Error"

### ISSUE THREAD

**Issue by dbkeitel** [ligovirgo/dqsegdb#40]
Last week, running any query, including this example from the aLIGO wiki
```
ligolw_segment_query_dqsegdb --segment-url https://segments.ligo.org --query-segments --include-segments L1:DMT-DC_READOUT_LOCKED --gps-start-time 1101600016 --gps-end-time `lalapps_tconvert now` | ligolw_print -t segment -c start_time -c end_time -d ' ' > segments.txt 
```
gave me the following error, which according to e-mail discussion was in the end due to my account on CIT having been very recently created. The purpose of this issue, as requested by Stuart Anderson and Ryan Fisher, is just the error handling and message reported to the user, the underlying issue in my case apparently was resolved from the admin side, or resolved itself, after a while.

Notes:

1. I got the same error on other machines I had long had accounts on, but I guess that's not important for the error-handling aspect.

1. running ligo-proxy-init before the query, the error turned into the more reasonable "Code 401 Unauthorized", but probably there could still be a more instructive message wrapped around that code.

1. The above example query has a rather long time span, which I just reproduce 1:1 so it matches the message quoted below. Increasing gps-start-time probably helps in debugging.

I think this specific error message I had saved is from ARCCA, but aside from the system path, it was the same on CIT:
```
/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/dqsegdb/urifunctions.py:339: UserWarning: Warning: Issue accessing url: https://segments.ligo.org/dq/L1/DMT-DC_READOUT_LOCKED
  warnings.warn("Warning: Issue accessing url: %s" % url)
/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/dqsegdb/urifunctions.py:340: UserWarning: Code: 500
  warnings.warn("Code: %s" % str(e.code))
/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/dqsegdb/urifunctions.py:341: UserWarning: Message: Internal Server Error
  warnings.warn("Message: %s" % str(e.msg))
/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/dqsegdb/urifunctions.py:344: UserWarning: May be handled cleanly by calling instance: otherwise will result in an error.
  warnings.warn("May be handled cleanly by calling instance: otherwise will result in an error.")
Traceback (most recent call last):
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/bin/ligolw_segment_query_dqsegdb", line 750, in <module>
    run_query_segments_dqsegdb(doc, process_id, protocol, server, gps_start_time, gps_end_time, options.include_segments, options.exclude_segments, options.result_name)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/bin/ligolw_segment_query_dqsegdb", line 525, in run_query_segments_dqsegdb

json_result,versioned_JSON,affected_results=apicalls.dqsegdbCascadedQuery(protocol, server, ifo, name, include_list_string, startTime, endTime)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/dqsegdb/apicalls.py", line 576, in dqsegdbCascadedQuery

result_flag,affected_results=clientutils.calculate_versionless_result(jsonResults,startTime,endTime,ifo_input=ifo)
UnboundLocalError: local variable 'jsonResults' referenced before assignment
Traceback (most recent call last):
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/bin/ligolw_print", line 211, in <module>
    xmldoc = utils.load_url(url, verbose = options.verbose, contenthandler = ContentHandler)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/glue/ligolw/utils/__init__.py", line 432, in load_url
    xmldoc, hexdigest = load_fileobj(fileobj, **kwargs)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/site-packages/glue/ligolw/utils/__init__.py", line 376, in load_fileobj
    ligolw.make_parser(contenthandler(xmldoc)).parse(fileobj)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/xml/sax/expatreader.py", line 110, in parse
    xmlreader.IncrementalParser.parse(self, source)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/xml/sax/xmlreader.py", line 125, in parse
    self.close()
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/xml/sax/expatreader.py", line 225, in close
    self.feed("", isFinal = 1)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/xml/sax/expatreader.py", line 217, in feed
    self._err_handler.fatalError(exc)
  File "/software/physics/ligo/spack/000/linux-redhat6-x86_64/gcc-5.4.0/ldg/ndjed33/lib/python2.7/xml/sax/handler.py", line 38, in fatalError
    raise exception
xml.sax._exceptions.SAXParseException: <unknown>:1:0: no element found 
```

**Comment by duncanmmacleod** [comment]
@dbkeitel, the second part of the error traceback is entirely due to `ligolw_print` not being able to parse what it expects as XML-format output from `ligolw_segment_query_dqsegdb`, because there is no XML output (it crashed).

That first command errors because of an authentication issue, but it looks like something else because of the bug reported in #12.

You can work around this bug by explicitly supplying a version number to your flag, i.e. query for `L1:DMT-DC_READOUT_LOCKED:1` instead of `L1:DMT-DC_READOUT_LOCKED`. I recommend doing this anyway, so that you know exactly what you will get back. On top of that you should request access through the proper channels (i.e. email @rpfisher).

Long story short, this report is a duplicate of #12, so please close the issue as long as my response makes sense.

**Comment by dbkeitel** [comment]
Hi Duncan, thanks for looking into it. As mentioned on DASWG (thread "ligolw_segment_query_dqsegdb: errors from https://segments.ligo.org") my personal issue with this was solved already on the authentication side, and I can successfully run queries. (I do indeed use the :1 in the actual queries I want to run, too.)

I don't understand nearly enough about this program to know if it's indeed a duplicate of #12, but as Ryan asked me explicitly to create an issue for the error handling, after the initial auth problem for me was solved, I propose waiting for him to have a look at this. Maybe I just didn't copy over all necessary information to this ticket to make it clear.

**Comment by duncanmmacleod** [comment]
@dbkeitel, there are three things happening at the moment `ligolw_segment_query_dqsegdb`:

1. when an unidentified user tries a query, the server returns a 500 exception, which is printed for the user, which isn't helpful
1. when an identified but unauthenticated user tries a query, the server returns a 401 exception, which is helpful(ish)
1. when an unauthenticated user tries a query and the server returns 401, the client ends up with an `UnboundLocalError` because of #12, which isn't helpful.

Item 3. is a duplicate of #12 as noted, 2. isn't a bug at all, but we could think about improving the error reporting from the client. 1. is a problem that needs to be resolved on the server side. I had originally only concentrated on the python exception being raised when I said this was a duplicate, but the 500 internal server error needs to be fixed, so this issue can act as the reminder for that.

**Comment by dbkeitel** [comment]
Ok thanks @duncanmmacleod , yes I (and presumably @rpfisher) meant this issue to address point 1! Feel free to clarify the title, or anything else to help tracking it. (I'm not all that familiar with github issues.)

**Comment by rpfisher** [comment]
Thanks for writing this up David.

-Ryan

-Sent from my Android phone

**Comment by rpfisher** [comment]
Resolved in afb057e04d0e6bd9c38c05d9a0d3e25785cbfbcb, now pushed to segments.ligo.org and segments-s6.ligo.org production servers.
